### PR TITLE
fix: support Ollama models for team members

### PR DESCRIPTION
**Pull request by 777genius**
## Summary

- replace static local-model rejection with runtime evidence for tools, context, and coordination
- preserve explicit Ollama and other custom local routes through project-scoped model selection
- add a one-shot experimental override for advisory coordination failures without bypassing provider, tool, 16K context, or real execution checks
- propagate the override through renderer, IPC, HTTP, controller, MCP, provisioning, and runtime launch
- add focused regression coverage and isolated local-provider preflight-to-launch E2E coverage
- extract launch helpers and lower legacy source-size caps

## Verification

- root focused suites: 631 tests passed
- agent-teams-controller: 169 tests passed
- agent-teams-mcp: 54 tests passed
- root and MCP typechecks passed
- source-size and team-provisioning architecture guards passed
- full fast source lint passed with 0 errors
- isolated local-provider live E2E passed in a temporary test project

## Summary by CodeRabbit

* **New Features**
  * Added an “experimental local model override” option to team creation and launch, with new localized label and hint text.
  * The `allowExperimentalLocalModels` flag is now supported end-to-end when enabled, affecting local-model override behavior in launches.

* **Bug Fixes**
  * Improved local-model coordination/readiness behavior with clearer warning vs blocking outcomes and more reliable retry/cancellation and bounded streaming handling.
  * Preserves explicitly selected local models in additional non-authoritative/incomplete-catalog scenarios.

* **Tests**
  * Expanded coverage for flag validation/propagation, UI override gating, and coordination-probe streaming and replay edge cases.

### CODE REVIEW

File: src/main/services/team/provisioning/TeamProvisioningMemberLifecycle.ts [hunk 1]
```diff
@@ -2534,6 +2532,7 @@ export class TeamProvisioningMemberLifecycleController {
     }
     this.assertRunStillCurrentAndAlive(run, teamName);
     const localModelPreflight = await adapter.preflightLocalModels?.({
+      allowExperimentalLocalModels: run.request.allowExperimentalLocalModels,
```
[the remark this answers is not in the document]
Comment (777genius) [reply src/main/services/team/provisioning/TeamProvisioningMemberLifecycle.ts:2535]: Fixed in 27c37fd90. The runtime owner now retains the consent, and the untracked pure OpenCode restart forwards it to both local preflight and adapter launch. Regression coverage exercises the completed-run restart path.

File: test/main/services/team/openCodeLiveTestHarness.ts [hunk 2]
```diff
@@ -449,8 +450,37 @@ async function assertExecutable(filePath: string): Promise<void> {
...
+      'Install the Bun version pinned by agent_teams_orchestrator/package.json before retrying.'
+  );
+}
+
 function withBunOnPath(pathValue: string): string {
-  const bunDir = '/Users/belief/.bun/bin';
+  const bunDir = path.join(os.userInfo().homedir, '.bun', 'bin');
   return pathValue.split(path.delimiter).includes(bunDir)
     ? pathValue
     : `${bunDir}${path.delimiter}${pathValue}`;
```
[the remark this answers is not in the document]
Comment (777genius) [reply test/main/services/team/openCodeLiveTestHarness.ts]: Fixed in 70b353a5b. The exact Bun directory that passes source-launcher preflight is now propagated into the bridge PATH, including BUN_INSTALL and system PATH candidates.

File: src/features/runtime-provider-management/main/infrastructure/OpenCodeLocalModelCoordinationProbe.ts [hunk 3]
```diff
@@ -253,6 +311,106 @@ function parseProbeResponse(raw: string): ProbeResponse | null {
...
+    if (!data || data === '[DONE]') continue;
+    try {
+      const parsed = asRecord(JSON.parse(data));
+      if (parsed) chunks.push(parsed);
+    } catch {
+      return [];
+    }
+  }
+  return chunks;
+}
```
[the remark this answers is not in the document]
Comment (777genius) [reply src/features/runtime-provider-management/main/infrastructure/OpenCodeLocalModelCoordinationProbe.ts:508]: Fixed in 9af5dd675. A malformed SSE frame is now skipped without discarding previously parsed chunks, and the Ollama streaming regression includes a malformed trailing frame.

File: src/renderer/components/team/dialogs/LaunchTeamDialog.tsx [hunk 4]
```diff
@@ -490,6 +397,7 @@ export const LaunchTeamDialog = (props: LaunchTeamDialogProps): React.JSX.Elemen
   const [prepareMessage, setPrepareMessage] = useState<string | null>(null);
   const [prepareWarnings, setPrepareWarnings] = useState<string[]>([]);
   const [prepareChecks, setPrepareChecks] = useState<ProvisioningProviderCheck[]>([]);
+  const [allowExperimentalLocalModels, setAllowExperimentalLocalModels] = useState(false);
```
[the remark this answers is not in the document]
Comment (777genius) [reply src/renderer/components/team/dialogs/LaunchTeamDialog.tsx:400]: Fixed in a32d0b3ee. LaunchTeamDialog now clears the one-shot consent in resetFormState before the dialog can be reopened.

File: src/renderer/components/team/dialogs/CreateTeamDialog.tsx [hunk 5]
```diff
@@ -1168,6 +1110,9 @@ export const CreateTeamDialog = ({
     () => buildProviderPrepareModelChecksSignature(selectedModelChecksByProvider),
     [selectedModelChecksByProvider]
   );
+  useEffect(() => {
+    setAllowExperimentalLocalModels(false);
+  }, [effectiveCwd, selectedModelChecksByProviderSignature]);
```
[the remark this answers is not in the document]
Comment (777genius) [reply src/renderer/components/team/dialogs/CreateTeamDialog.tsx:1115]: Fixed in a32d0b3ee. CreateTeamDialog now clears the one-shot consent in resetUIState when the dialog is dismissed or completes.

**Comment by 777genius** [comment]
Review follow-up:

- Addressed and resolved all 5 inline review threads.
- Added regressions for completed-run restart consent and malformed Ollama SSE tails.
- Kept deep prepare intentionally strict. That pass discovers the blocking local-model issue and exposes the one-shot consent checkbox. The confirmed override is applied only to the create or launch request and definitive runtime preflight. Passing it into prepare before consent would bypass the product safety boundary.
- Local verification: 630 provisioning tests plus 69 focused review tests passed, typecheck passed, type-aware lint has 0 errors, and both source-size and Team Provisioning architecture guards passed.

@codex review

**Comment by 777genius** [comment]
Addressed the incremental streaming identity finding in ef60a5703. Tool-call fragments now group by stable id first, use index only while it remains unambiguous, and keep distinct calls separate when Ollama repeats an index. Fixtures cover omitted indexes, repeated indexes, split arguments, and malformed trailing SSE frames.

Verification: 8/8 focused tests, typecheck, type-aware lint with 0 errors, and source-size guard passed.

@codex review

File: src/main/services/team/provisioning/TeamProvisioningOpenCodeRuntimeAdapterLaunch.ts [hunk 6]
```diff
@@ -151,6 +154,9 @@ export function buildOpenCodeRuntimeAdapterLaunchInput(
       model: params.request.model,
       effort: params.request.effort,
       skipPermissions: params.request.skipPermissions !== false,
+      ...(params.request.allowExperimentalLocalModels === true
+        ? { allowExperimentalLocalModels: true }
+        : {}),
```
[the remark this answers is not in the document]
Comment (777genius) [reply src/main/services/team/provisioning/TeamProvisioningOpenCodeRuntimeAdapterLaunch.ts:161]: Fixed in 721353d14. The pure OpenCode launch owner now stores allowExperimentalLocalModels, and aggregate cleanup owner reconstructions preserve the same run consent as well. Regression coverage verifies the successful launch owner keeps the flag.

File: src/main/services/team/provisioning/TeamProvisioningOpenCodeModelPreparation.ts [hunk 7]
```diff
@@ -498,14 +518,27 @@ async function prepareSelectedOpenCodeModelsCompatibilityBatch({
...
   let sharedPrepare: TeamRuntimePrepareResult;
   try {
     sharedPrepare = await adapter.prepare({
       runId: `prepare-${randomUUID()}`,
       teamName: '__prepare_opencode__',
       cwd,
       providerId: 'opencode',
-      model: undefined,
+      model: catalogModelIds[0],
       runtimeOnly: true,
```
[the remark this answers is not in the document]
Comment (777genius) [reply src/main/services/team/provisioning/TeamProvisioningOpenCodeModelPreparation.ts]: Fixed in 91d69505f. Compatibility preparation now classifies app-configured custom local routes through the local provider inventory before shared OpenCode authentication. Classification does not run the coordination probe; deep verification remains responsible for runtime/model capability checks. A fallback also prevents an unknown provider route from short-circuiting before catalog deferral when classification is unavailable.

File: src/main/services/team/provisioning/TeamProvisioningMemberLifecycle.ts [hunk 8]
```diff
@@ -2051,6 +2047,7 @@ export class TeamProvisioningMemberLifecycleController {
     }
 
     const localModelPreflight = await adapter.preflightLocalModels?.({
+      allowExperimentalLocalModels: runtimeRun.allowExperimentalLocalModels,
```
[the remark this answers is not in the document]
Comment (777genius) [reply src/main/services/team/provisioning/TeamProvisioningMemberLifecycle.ts:2050]: Fixed in 721353d14. Runtime approval owner refreshes now copy allowExperimentalLocalModels from the exact captured owner, with a regression test covering the approval-to-restart path.

**Comment by 777genius** [comment]
@codex review

**Comment by 777genius** [comment]
@codex review

**Comment by 777genius** [comment]
@codex review

File: src/main/services/team/provisioning/TeamProvisioningOpenCodeModelPreparation.ts [hunk 9]
```diff
@@ -498,29 +532,62 @@ async function prepareSelectedOpenCodeModelsCompatibilityBatch({
...
-      providerId: 'opencode',
-      reason: 'unknown_error',
-      retryable: false,
-      diagnostics: [message],
-      warnings: [],
-    };
+    return { details, warnings, blockingMessages, issues, supportDiagnostics };
+  }
+
+  let sharedPrepare = await prepareOpenCodeCompatibilityModel(adapter, cwd, catalogModelIds[0]);
```
[the remark this answers is not in the document]
Comment (777genius) [reply src/main/services/team/provisioning/TeamProvisioningOpenCodeModelPreparation.ts:574]: Fixed in 5c9c9b817. Cursor ACP and Kiro routes are now deferred directly to deep execution verification, like configured local routes. When they are mixed with provider-backed models, the shared runtime-only auth check selects the provider-backed route instead. Regression coverage includes Cursor, Kiro, Ollama, and a mixed Cursor plus OpenRouter selection.

**Comment by 777genius** [comment]
@codex review

**Comment by 777genius** [comment]
@codex review

**Comment by 777genius** [comment]
@codex review

File: src/features/runtime-provider-management/main/infrastructure/OpenCodeLocalModelRuntimeInspector.ts [hunk 10]
```diff
@@ -473,15 +526,29 @@ function buildCoordinationProbeFailure(input: {
     configuredContextTokens: input.configuredContextTokens ?? null,
     effectiveContextTokens: input.effectiveContextTokens ?? null,
     coordinationProbeStatus: input.coordination.status,
-    severity: 'blocking',
-    code:
-      input.coordination.status === 'failed'
-        ? 'local_coordination_probe_failed'
-        : 'local_coordination_probe_unavailable',
-    message: input.coordination.message,
+    severity: unavailable || overrideApplied ? 'warning' : 'blocking',
```
[the remark this answers is not in the document]
Comment (777genius) [reply src/features/runtime-provider-management/main/infrastructure/OpenCodeLocalModelRuntimeInspector.ts:531]: Fixed in 374534bde. HTTP 400/422 client rejections from the coordination request are now classified as failed rather than unavailable, so they stay blocking unless the user explicitly enables the experimental override. Transport, timeout, rate-limit, and server availability failures remain retryable warnings. Regression coverage keeps HTTP 503 unavailable and HTTP 400 tool rejection failed.

**Comment by 777genius** [comment]
@codex review

**Comment by 777genius** [comment]
@codex review

**Comment by 777genius** [comment]
@codex review

**Comment by 777genius** [comment]
@codex review

**Comment by 777genius** [comment]
@codex review

**Comment by 777genius** [comment]
@codex review

File: src/features/runtime-provider-management/main/infrastructure/OpenCodeLocalModelRuntimeInspector.ts [hunk 11]
```diff
@@ -462,7 +511,11 @@ function buildCoordinationProbeFailure(input: {
   configuredContextTokens?: number | null;
   effectiveContextTokens?: number | null;
   coordination: OpenCodeLocalModelCoordinationProbeResult;
+  allowExperimentalLocalModels: boolean;
 }): OpenCodeLocalModelRuntimeReadiness {
+  const unavailable = input.coordination.status === 'unavailable';
+  const experimentalOverride = input.coordination.status === 'failed';
+  const overrideApplied = experimentalOverride && input.allowExperimentalLocalModels;
```
[the remark this answers is not in the document]
Comment (777genius) [reply src/features/runtime-provider-management/main/infrastructure/OpenCodeLocalModelRuntimeInspector.ts]: Fixed in c5a485e30. The coordination probe now preserves request_rejected as explicit evidence. HTTP 400/422 rejection stays blocking even when the experimental override is enabled, is reported as local_coordination_probe_failed, and does not advertise the override. Behavioral coordination failures remain overrideable. Regression coverage verifies both paths.

**Comment by 777genius** [comment]
@codex review\n\nAlso fixed the ReviewRouter lookup finding in ba9ef3f2b: when the local-provider lookup is non-authoritative, saved custom routes are deferred to deep preflight instead of being rejected by catalog-only renderer validation. Authoritatively unknown routes remain blocked.

**Comment by 777genius** [comment]
@codex review

File: src/renderer/components/team/dialogs/useOpenCodeLocalModelScope.ts [hunk 12]
```diff
@@ -0,0 +1,40 @@
...
+  projectPath: string;
+  selectedProviderId: TeamProviderId;
+  members: readonly MemberProviderSelection[];
+}): OpenCodeLocalModelScope {
+  const requiresLookup =
+    input.selectedProviderId === 'opencode' ||
+    input.members.some((member) => !member.removedAt && member.providerId === 'opencode');
+  const { providers, authoritative } = useOpenCodeLocalProviders({
+    enabled: input.enabled && requiresLookup,
+    projectPath: input.projectPath || null,
```
[the remark this answers is not in the document]
Comment (777genius) [reply src/renderer/components/team/dialogs/useOpenCodeLocalModelScope.ts]: Fixed in 30bf7c3fe. The member local-provider scope lookup now stays disabled and non-authoritative until a non-empty project path is resolved. This prevents a global-only snapshot from clearing saved project-scoped teammate models during dialog hydration. Focused tests cover the unresolved path, project lookup enablement, and preservation of inherited custom routes.

**Comment by 777genius** [comment]
@codex review
